#' Extract codon information at E, P, and A sites for each read from ribo-seq data.
#'
#' @param cdna_path Corresponding species cDNA.fa file
#' @param reads_psite_list List file generated by the psite_info function of riboWaltz
#' @param geneinfo The output of geneinfo, a Genome annotation matrix
#' @param tx_scale If transcripts were not filtered in the previous steps, here the transcript with the
#'    longest CDS will be used to represent the gene.If it is already one gene to one transcript, tx_scale = "no".
#'    default is "the_longest"
#'
#' @examples
#'
#' ## EPA_result<-EPA_res(cdna_path = "../Mus_musculus.GRCm39.cdna.all.fa",reads_psite_list = reads_psite_list,geneinfo,tx_scale ="no")
#'
#' @export
#'
EPA_res <- function(cdna_path, reads_psite_list, geneinfo, tx_scale = "the_longest") {
  EPA_res_list <- list()
  sequences <- Biostrings::readDNAStringSet(cdna_path, format = "fasta", use.names = TRUE)
  names(sequences) <- sub("\\..*", "", names(sequences))

  geneinfo <- geneinfo[which(geneinfo$transcript_biotype == "protein_coding"), ]
  geneinfo <- geneinfo[-which(geneinfo$chrom == "MT"), ]
  if (tx_scale == "the_longest") {
    geneinfo <- geneinfo[order(geneinfo$cds_len, decreasing = T), ]
    geneinfo <- geneinfo[!duplicated(geneinfo$gene_id), ]
  } else if (tx_scale == "no") {
    geneinfo <- geneinfo[order(geneinfo$cds_len, decreasing = T), ]
  }
  for (i in seq_len(length(reads_psite_list))) {
    A <- as.data.frame(reads_psite_list[[i]])
    A <- A[which(A[, 10] == "cds"), ]
    A <- A[which(A[, 8] %% 3 == 0), ]
    A <- A[, c(1, 3)]

    EPA_res <- A[which(A[, 1] %in% geneinfo$tx_name), ]
    colnames(EPA_res) <- c("tx_name", "trans.site")
    EPA_res <- EPA_res[which(EPA_res$tx_name %in% names(sequences)), ]
    if (any(EPA_res$trans.site < 4)) {
      EPA_res<-EPA_res[-which(EPA_res$trans.site < 4),]
    }
    EPA_res$E_Codon <- as.character(Biostrings::subseq(sequences[as.character(EPA_res$tx_name)],
      start = EPA_res$trans.site - 3,
      end = EPA_res$trans.site - 1
    ))
    EPA_res$P_Codon <- as.character(Biostrings::subseq(sequences[as.character(EPA_res$tx_name)],
      start = EPA_res$trans.site,
      end = EPA_res$trans.site + 2
    ))
    EPA_res$A_Codon <- as.character(Biostrings::subseq(sequences[as.character(EPA_res$tx_name)],
      start = EPA_res$trans.site + 3,
      end = EPA_res$trans.site + 5
    ))

    name <- paste("res", as.character(names(reads_psite_list[i])), sep = "_")
    EPA_res_list[[name]] <- EPA_res
  }
  return(EPA_res_list)
}
